kv/kvserver: TestStoreRangeSplitBackpressureWrites failed

kv/kvserver.TestStoreRangeSplitBackpressureWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21346325?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21346325?buildTab=artifacts#/) on release-25.4 @ [21dc029f057bd91b3878362bf1f66622bf7570df](https://github.com/cockroachdb/cockroach/commits/21dc029f057bd91b3878362bf1f66622bf7570df):


```
           102.593ms      0.011ms                    event:kv/kvserver/replica_write.go:432 [n1,s1,r79/1:/{Table/100-Max}] executing read-write batch
           102.692ms      0.098ms                    event:kv/kvserver/batcheval/cmd_end_transaction.go:275 [n1,s1,r79/1:/{Table/100-Max}] checking to see if transaction record already exists for txn: "unnamed" meta={id=c1daba1f key=/Table/100 iso=Serializable pri=0.02235784 epo=0 ts=1778726365.374534190,0 min=1778726365.374534190,0 seq=3} lock=true stat=PENDING rts=1778726365.374534190,0 gul=1778726365.874534190,0 obs={n1@1778726365.374534190,0}
           102.773ms      0.081ms                    event:kv/kvserver/batcheval/cmd_end_transaction.go:286 [n1,s1,r79/1:/{Table/100-Max}] no existing txn record found
           103.147ms      0.374ms                    event:kv/kvserver/replica_evaluate.go:567 [n1,s1,r79/1:/{Table/100-Max}] evaluated EndTxn command header:<key:"\354" sequence:3 > deadline:<> lock_spans:<key:"\354" > , txn="unnamed" meta={id=c1daba1f key=/Table/100 iso=Serializable pri=0.02235784 epo=0 ts=1778726365.374534190,0 min=1778726365.374534190,0 seq=3} lock=true stat=PENDING rts=1778726365.374534190,0 gul=1778726365.874534190,0 obs={n1@1778726365.374534190,0} : resp=header:<txn:<meta:<id:c1daba1f-a6c6-491c-9577-720740a09ab5 key:"\354" write_timestamp:<wall_time:1778726365374534190 > min_timestamp:<wall_time:1778726365374534190 > priority:480131 sequence:3 coordinator_node_id:1 > name:"unnamed" status:ABORTED last_h..., err=<nil>
           103.163ms      0.016ms                    event:kv/kvserver/replica_raft.go:121 [n1,s1,r79/1:/{Table/100-Max}] evaluated request
           103.185ms      0.619ms                event:server/node.go:1728 [n1] node sending response
           103.210ms      0.709ms            event:kv/kvclient/kvcoord/transport.go:212 [n1,txn=c1daba1f] received batch response
    client_split_test.go:1825: TRACE FOR put:
             0.000ms      0.000ms    === operation:put gid:39950 _verbose:1
             0.000ms      0.000ms    [dist sender send: {count: 1, duration 327ms}]
             0.000ms      0.000ms    [/cockroach.roachpb.Internal/Batch: {count: 1, duration 327ms}]
             0.000ms      0.000ms    [executeWriteBatch: {count: 1, duration 665µs}]
             0.000ms      0.000ms    [local proposal: {count: 1, duration 88µs}]
             0.048ms      0.048ms        === operation:dist sender send gid:42402 _verbose:1 node:1
             0.048ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 1, duration 327ms}]
             0.048ms      0.000ms        [executeWriteBatch: {count: 1, duration 665µs}]
             0.048ms      0.000ms        [local proposal: {count: 1, duration 88µs}]
             0.106ms      0.058ms        event:kv/kvclient/kvcoord/range_iter.go:183 [n1] querying next range at /Table/100
             0.140ms      0.035ms        event:kv/kvclient/kvcoord/range_iter.go:220 [n1] key: /Table/100, desc: r79:/{Table/100-Max} [(n1,s1):1, next=2, gen=1]
             0.160ms      0.019ms        event:kv/kvclient/kvcoord/dist_sender.go:2726 [n1] r79: sending batch 1 Put to (n1,s1):1
             0.175ms      0.015ms        event:rpc/nodedialer/nodedialer.go:167 [n1] sending request to local client
             0.185ms      0.010ms        event:kv/kvclient/kvcoord/transport.go:210 [n1] sending batch request
             0.189ms      0.004ms            === operation:/cockroach.roachpb.Internal/Batch gid:42402 _verbose:1 node:1 span.kind:server request:Put [/Table/100]
             0.189ms      0.000ms            [executeWriteBatch: {count: 1, duration 665µs}]
             0.189ms      0.000ms            [local proposal: {count: 1, duration 88µs}]
             0.199ms      0.010ms            event:server/node.go:1623 [n1] node received request: 1 Put
             0.261ms      0.061ms            event:kv/kvserver/store_send.go:149 [n1,s1] executing Put [/Table/100]
             0.284ms      0.023ms            event:kv/kvserver/replica_backpressure.go:206 [n1,s1,r79/1:/{Table/100-Max}] applying backpressure to limit range growth on batch Put [/Table/100]
           325.978ms    325.694ms            event:kv/kvserver/replica_send.go:187 [n1,s1,r79/1:/{Table/100-Max}] read-write path
           325.998ms      0.019ms            event:kv/kvserver/concurrency/concurrency_manager.go:264 [n1,s1,r79/1:/{Table/100-Max}] sequencing request
           326.014ms      0.016ms            event:kv/kvserver/concurrency/concurrency_manager.go:345 [n1,s1,r79/1:/{Table/100-Max}] acquiring latches
           326.031ms      0.017ms            event:kv/kvserver/concurrency/concurrency_manager.go:389 [n1,s1,r79/1:/{Table/100-Max}] scanning lock table for conflicting locks
           326.035ms      0.004ms                === operation:executeWriteBatch gid:42402 _verbose:1 node:1 store:1 range:79/1:/{Table/100-Max}
           326.035ms      0.000ms                [local proposal: {count: 1, duration 88µs}]
           326.067ms      0.032ms                event:kv/kvserver/replica_write.go:176 [n1,s1,r79/1:/{Table/100-Max}] applied timestamp cache
           326.087ms      0.020ms                event:kv/kvserver/replica_write.go:432 [n1,s1,r79/1:/{Table/100-Max}] executing read-write batch
           326.410ms      0.323ms                event:kv/kvserver/replica_evaluate.go:567 [n1,s1,r79/1:/{Table/100-Max}] evaluated Put command header:<key:"\354" > value:<raw_bytes:"n\337\241\232\003test" timestamp:<> > expect_exclusion_since:<> , txn=<nil> : resp=header:<> , err=<nil>
           326.425ms      0.015ms                event:kv/kvserver/replica_proposal.go:1050 [n1,s1,r79/1:/{Table/100-Max}] need consensus on write batch with op count=1
           326.442ms      0.018ms                event:kv/kvserver/replica_raft.go:121 [n1,s1,r79/1:/{Table/100-Max}] evaluated request
           326.461ms      0.018ms                event:kv/kvserver/replica_raft.go:176 [n1,s1,r79/1:/{Table/100-Max}] proposing command to write 1 new keys, 1 new values, 0 new intents, write batch size=37 bytes
           326.480ms      0.019ms                event:kv/kvserver/replica_raft.go:290 [n1,s1,r79/1:/{Table/100-Max}] acquiring proposal quota (194 bytes)
           326.497ms      0.018ms                event:kv/kvserver/replica_raft.go:479 [n1,s1,r79/1:/{Table/100-Max}] submitting proposal to proposal buffer
           326.529ms      0.032ms                event:kv/kvserver/replica_proposal_buf.go:561 [n1,s1,r79/1:/{Table/100-Max}] flushing proposal to Raft
           326.551ms      0.022ms                event:kv/kvserver/rafttrace/rafttrace.go:318 [n1,s1,r79/1:/{Table/100-Max}] too many active raft traces, skipping
           326.602ms      0.052ms                    === operation:local proposal gid:40734 _verbose:1 node:1 store:1 range:79/1:/{Table/100-Max} raft:
           326.630ms      0.028ms                    event:kv/kvserver/app_batch.go:120 [n1,s1,r79/1:/{Table/100-Max},raft] applying command
           326.684ms      0.054ms                    event:kv/kvserver/replica_application_state_machine.go:188 [n1,s1,r79/1:/{Table/100-Max},raft] LocalResult (reply: (err: <nil>), *kvpb.PutResponse, #encountered intents: 0, #acquired locks: 0, #resolved locks: 0#updated txns: 0 #end txns: 0, PopulateBarrierResponse:false RepopulateSubsumeResponse:false GossipFirstRange:false MaybeGossipSystemConfig:false MaybeGossipSystemConfigIfHaveFailure:false MaybeAddToSplitQueue:false MaybeGossipNodeLiveness:<nil> 
           326.720ms      0.689ms            event:server/node.go:1728 [n1] node sending response
           326.745ms    326.561ms        event:kv/kvclient/kvcoord/transport.go:212 [n1] received batch response
--- FAIL: TestStoreRangeSplitBackpressureWrites/splitOngoing=true,splitErr=true (0.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeSplitBackpressureWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63927

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -1584,7 +1584,16 @@ func TestStoreRangeSplitBackpressureWrites(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			var activateSplitFilter int32
 			splitKey := roachpb.RKey(bootstrap.TestingUserTableDataMin(keys.SystemSQLCodec))
-			splitPending, blockSplits := make(chan struct{}), make(chan struct{})
+			// splitPending is buffered so the request filter doesn't block
+			// indefinitely if the test fails before reading the signal.
+			splitPending, blockSplits := make(chan struct{}, 1), make(chan struct{})
+			// backpressureRegistered receives one signal per write that
+			// successfully registered a backpressure callback with the split
+			// queue. This lets the test wait until both writes are registered
+			// before unblocking the split, avoiding a time-based coordination
+			// race on slow hardware (e.g. s390x). Buffered so the knob never
+			// blocks if the test stops listening early.
+			backpressureRegistered := make(chan roachpb.RangeID, 8)
 
 			// Set maxBytes to something small so we can exceed the maximum split
 			// size without adding 2x64MB of data.
@@ -1626,6 +1635,12 @@ func TestStoreRangeSplitBackpressureWrites(t *testing.T) {
 						DisableMergeQueue:              true,
 						DisableSplitQueue:              true,
 						TestingRequestFilter:           testingRequestFilter,
+						TestingBackpressureCallbackRegistered: func(id roachpb.RangeID) {
+							select {
+							case backpressureRegistered <- id:
+							default:
+							}
+						},
 					},
 				},
 			})
@@ -1664,7 +1679,11 @@ func TestStoreRangeSplitBackpressureWrites(t *testing.T) {
 				}); err != nil {
 					t.Fatal(err)
 				}
-				<-splitPending
+				select {
+				case <-splitPending:
+				case <-time.After(testutils.DefaultSucceedsSoonDuration):
+					t.Fatal("timed out waiting for split filter to trigger")
+				}
 			} else if tc.splitImpossible {
 				store.TestingSetSplitQueueActive(true)
 				if err := store.ForceSplitScanAndProcess(); err != nil {
@@ -1701,17 +1720,29 @@ func TestStoreRangeSplitBackpressureWrites(t *testing.T) {
 				})
 			}()
 
-			// Make sure the write doesn't return while a split is ongoing. If no
-			// split is ongoing, the write will return an error immediately.
+			// Wait for both writes to register a backpressure callback before
+			// unblocking the split. This ensures both writes will observe the
+			// split's outcome via their registered callbacks, avoiding a race
+			// where the split completes (and removes the range from the split
+			// queue's tracked replicas) before a slow goroutine calls
+			// MaybeAddCallback. While waiting, also fail fast if either write
+			// returns prematurely.
 			if tc.splitOngoing {
-				select {
-				case err := <-putRes:
-					close(blockSplits)
-					t.Fatalf("put was not blocked on split, returned err %v", err)
-				case err := <-delRes:
-					close(blockSplits)
-					t.Fatalf("delete was not blocked on split, returned err %v", err)
-				case <-time.After(100 * time.Millisecond):
+				timeout := time.After(testutils.DefaultSucceedsSoonDuration)
+				for registered := 0; registered < 2; {
+					select {
+					case err := <-putRes:
+						close(blockSplits)
+						t.Fatalf("put was not blocked on split, returned err %v", err)
+					case err := <-delRes:
+						close(blockSplits)
+						t.Fatalf("delete was not blocked on split, returned err %v", err)
+					case <-backpressureRegistered:
+						registered++
+					case <-timeout:
+						close(blockSplits)
+						t.Fatalf("timed out waiting for backpressure callbacks to register (got %d/2)", registered)
+					}
 				}
 
 				// Let split through. Write should follow.
@@ -1723,7 +1754,13 @@ func TestStoreRangeSplitBackpressureWrites(t *testing.T) {
 				"put":    putRes,
 				"delete": delRes,
 			} {
-				if err := <-resCh; tc.expErr == "" {
+				var err error
+				select {
+				case err = <-resCh:
+				case <-time.After(testutils.DefaultSucceedsSoonDuration):
+					t.Fatalf("timed out waiting for %s to return", op)
+				}
+				if tc.expErr == "" {
 					if err != nil {
 						t.Fatalf("%s returned err %v, expected success", op, err)
 					}

--- a/pkg/kv/kvserver/replica_backpressure.go
+++ b/pkg/kv/kvserver/replica_backpressure.go
@@ -241,6 +241,10 @@ func (r *Replica) maybeBackpressureBatch(ctx context.Context, ba *kvpb.BatchRequ
 			return nil
 		}
 
+		if knobs := r.store.TestingKnobs(); knobs != nil && knobs.TestingBackpressureCallbackRegistered != nil {
+			knobs.TestingBackpressureCallbackRegistered(r.RangeID)
+		}
+
 		const errHint = `For help understanding this error and troubleshooting, visit:
 
     https://www.cockroachlabs.com/docs/stable/common-errors.html#split-failed-while-applying-backpressure-are-rows-updated-in-a-tight-loop`

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -65,6 +65,14 @@ type StoreTestingKnobs struct {
 	// handled and the batch is retried.
 	TestingConcurrencyRetryFilter kvserverbase.ReplicaConcurrencyRetryFilter
 
+	// TestingBackpressureCallbackRegistered, if set, is called from
+	// (*Replica).maybeBackpressureBatch after a backpressure callback has been
+	// successfully registered with the split queue. Tests use this to
+	// deterministically learn that a backpressured write is waiting on the
+	// callback (and thus will observe the split's outcome), which avoids
+	// time-based coordination on slow hardware.
+	TestingBackpressureCallbackRegistered func(roachpb.RangeID)
+
 	// TestingProposalFilter is called before proposing each command.
 	TestingProposalFilter kvserverbase.ReplicaProposalFilter
 	// TestingProposalSubmitFilter can be used by tests to observe and optionally

